<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>1. Introduction &#8212; Platypus 0.1.5 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.5',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Welcome to Platypus 0.1.5 documentation!" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="index.html" title="Welcome to Platypus 0.1.5 documentation!"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Platypus 0.1.5 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="introduction">
<h1>1. Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h1>
<p>Platypus is a tool for identifying small variants (SNPs and indels) in BAM files. This code is a fork of Platypus version 0.1.5, which is used
in OpEx. If you are looking for the most recent version of Platypus, the code is available <a class="reference external" href="https://github.com/andyrimmer/Platypus">here</a>. Platypus
uses a haplotype-based approach to variant calling, which enables fast and accurate short indel calling.</p>
</div>
<div class="section" id="installation">
<h1>2. Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p>Stable releases of Platypus can be downloaded from Github <a class="reference external" href="https://github.com/RahmanTeamDevelopment/Platypus/releases">here</a>
in either <code class="docutils literal"><span class="pre">.zip</span></code> or <code class="docutils literal"><span class="pre">.tar.gz</span></code> format. To unpack these run one of the following commands:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">unzip</span> <span class="n">Platypus</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">zip</span>
</pre></div>
</div>
<p>or:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">tar</span> <span class="o">-</span><span class="n">xvzf</span> <span class="n">Platypus</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>and then you can install Platypus with the following commands:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="n">Platypus</span>
<span class="o">./</span><span class="n">install</span><span class="o">.</span><span class="n">sh</span>
</pre></div>
</div>
<p>Platypus uses <code class="docutils literal"><span class="pre">virtualenv</span></code> and <code class="docutils literal"><span class="pre">pip</span></code> to manage all its extra dependencies, which means that it will not clutter up your system by installing
things globally. Everthing it installs will go into a sub-directory in the <code class="docutils literal"><span class="pre">Platypus</span></code> directory (specifically, <code class="docutils literal"><span class="pre">Platypus_X.X.X/env</span></code>). If
you delete Platypus then everything it has installed will also be deleted. Once the installation script has finished successfully,
Platypus is ready for use.</p>
<div class="section" id="dependencies">
<h2>2.1. Dependencies<a class="headerlink" href="#dependencies" title="Permalink to this headline">¶</a></h2>
<p>To install and run Platypus v0.1.5 you will need <a class="reference external" href="https://www.python.org">Python</a> (only version 2.7.X is currently supported),
and <a class="reference external" href="https://virtualenv.pypa.io/en/stable/">Virtualenv</a>.</p>
</div>
</div>
<div class="section" id="running-platypus">
<h1>3. Running Platypus<a class="headerlink" href="#running-platypus" title="Permalink to this headline">¶</a></h1>
<p>Once downloaded and correctly installed, Platypus can be run with the following command:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">env</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">platypus</span> <span class="n">callVariants</span> <span class="o">--</span><span class="n">bamFiles</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">bam</span> <span class="o">--</span><span class="n">refFile</span><span class="o">=</span><span class="n">reference</span><span class="o">.</span><span class="n">fa</span> <span class="o">--</span><span class="n">output</span><span class="o">=</span><span class="n">variant_calls</span><span class="o">.</span><span class="n">vcf</span>
</pre></div>
</div>
<p>This is the simplest way to run Platypus, specifying only the names of the input BAM file and input reference files, and the output VCF file. Variant
calls will then be outupt to a file called <cite>variant_calls.vcf</cite>.</p>
<ul class="simple">
<li>The input bam file should follow the <a class="reference external" href="http://samtools.github.io/hts-specs/SAMv1.pdf">BAM format</a> and should contain mapped reads sorted by genomic coordinate. Platypus has been tested using BAM files produced by <a class="reference external" href="http://www.well.ox.ac.uk/project-stampy">Stampy</a> and <a class="reference external" href="http://bio-bwa.sourceforge.net">BWA</a>, but should also work with other short read aligners. The BAM index file (.bai) should also be present in the same directory; this can be created using <a class="reference external" href="http://www.htslib.org">Samtools</a>.</li>
</ul>
</div>
<div class="section" id="command-line-options">
<h1>4. Command Line Options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h1>
<p>Platypus has a large number of configuration options that can be specified on the command-line. Most of these can be ignored, as they take sensible default values
and are already configured for optimal results. Below is a full descripton of all configuration options.</p>
<div class="section" id="basic-options">
<h2>4.1. Basic options<a class="headerlink" href="#basic-options" title="Permalink to this headline">¶</a></h2>
<table border="1" class="docutils">
<colgroup>
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Option</th>
<th class="head">Type</th>
<th class="head">Default Value</th>
<th class="head">Example Usage</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>-h &#8211;help</td>
<td>NA</td>
<td>NA</td>
<td><cite>&#8211;help</cite></td>
<td>If this option is set then Platypus will print a help message and then exit</td>
</tr>
<tr class="row-odd"><td>&#8211;bamFiles</td>
<td>Comma-separated list of strings</td>
<td>None</td>
<td><cite>&#8211;bamFiles=test1.bam,test2.bam</cite></td>
<td>A list of BAM file names. All the specified BAM files will be searched for variants. The output will go into a single VCF file</td>
</tr>
<tr class="row-even"><td>&#8211;refFile</td>
<td>String</td>
<td>None</td>
<td><cite>&#8211;refFile=human_37.fa</cite></td>
<td>Name of the input reference FASTA file. This file must be indexed (using <cite>Samtools faidx</cite> for example) and must contain the same reference sequence that was used to map the reads in the input BAM file(s)</td>
</tr>
<tr class="row-odd"><td>-o &#8211;output</td>
<td>String</td>
<td>AllVariants.vcf</td>
<td><cite>&#8211;output=variant_calls.vcf</cite></td>
<td>Name of the output VCF file</td>
</tr>
<tr class="row-even"><td>&#8211;regions</td>
<td>Comma-separated list of strings or the name of a text file</td>
<td>None</td>
<td><cite>&#8211;regions=chr1:1000-2000,chr2:2000-4000</cite></td>
<td>Platypus will produce variant calls only in the specified regions. See <a class="reference internal" href="#specifying-calling-regions"><span class="std std-ref">Specifying calling regions</span></a> for a more detailed description</td>
</tr>
<tr class="row-odd"><td>&#8211;verbosity</td>
<td>Integer from 0 to 3</td>
<td>2</td>
<td><cite>&#8211;verbosity=3</cite></td>
<td>Sets the level of output logging for Platypus. Increase this value for more verbose log output</td>
</tr>
<tr class="row-even"><td>&#8211;logFileName</td>
<td>String</td>
<td>log.txt</td>
<td><cite>&#8211;logFileName=platypus_log_file.txt</cite></td>
<td>Name of the output log file.</td>
</tr>
<tr class="row-odd"><td>&#8211;nCPU</td>
<td>Integer</td>
<td>1</td>
<td><cite>&#8211;nCPU=3</cite></td>
<td>Number of processes to use for variant calling. If &gt; 1 then Platypus will run in multiple processes and merge the VCF files into one file at the end</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="bam-data-processing-and-filtering-options">
<span id="bam-data-quality-filtering"></span><h2>4.2. BAM Data processing and filtering options<a class="headerlink" href="#bam-data-processing-and-filtering-options" title="Permalink to this headline">¶</a></h2>
<p>Platypus does not necessarily use every read, in the BAM for variant calling. Whole reads may be filtered out based on various quality criteria, and segments of individual reads may also be filtered. The table below gives a summary of the configuration options that may be used to control read-level filtering.</p>
<table border="1" class="docutils">
<colgroup>
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Option</th>
<th class="head">Type</th>
<th class="head">Default Value</th>
<th class="head">Example Usage</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&#8211;processRegionSize</td>
<td>Integer</td>
<td>30,000,000</td>
<td><cite>&#8211;processRegionSize=10000000</cite></td>
<td>Platypus breaks up the genome into regions of this size and processes them in parallel (if nCPU &gt; 1) or consecutively (if nCPU == 1)</td>
</tr>
<tr class="row-odd"><td>&#8211;bufferSize</td>
<td>Integer</td>
<td>1,000,000</td>
<td><cite>&#8211;bufferSize=100000</cite></td>
<td>The maximum size of region (per process, and as a genomic interval) that Platypus will read into memory at any point. This can be used to control memory usage</td>
</tr>
<tr class="row-even"><td>&#8211;maxReads</td>
<td>Integer</td>
<td>5,000,000</td>
<td><cite>&#8211;maxReads=100000</cite></td>
<td>This sets an upper limit on the amount of data that Platypus will try to process in one region. If the number of reads in a region of <cite>bufferSize</cite> is larger than this then Platypus will skip the region. This option can be used, in conjunction with <cite>bufferSize</cite> to control memory usage.</td>
</tr>
<tr class="row-odd"><td>&#8211;maxBadQualBases</td>
<td>Integer</td>
<td>20</td>
<td><cite>&#8211;maxBadQualBases=10</cite></td>
<td>If a read has &gt; this number of bases with quality &lt; 20 then it will be filtered out and not used for variant calling.</td>
</tr>
<tr class="row-even"><td>&#8211;minGoodQualBases</td>
<td>Integer</td>
<td>20</td>
<td><cite>&#8211;minGoodQualBases=10</cite></td>
<td>If a read has &lt; this number of bases with quality &gt;= 20 then it will be filtered out and not used for variant calling.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="variant-calling-options">
<h2>4.3. Variant calling options<a class="headerlink" href="#variant-calling-options" title="Permalink to this headline">¶</a></h2>
<table border="1" class="docutils">
<colgroup>
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Option</th>
<th class="head">Type</th>
<th class="head">Default Value</th>
<th class="head">Example Usage</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&#8211;genSNPs</td>
<td>Boolean</td>
<td>1 (True)</td>
<td><cite>&#8211;genSNPs=1</cite></td>
<td>If set to 1, Platypus will call SNPs. If set to 0 Platypus will not call SNPs</td>
</tr>
<tr class="row-odd"><td>&#8211;genIndels</td>
<td>Boolean</td>
<td>1 (True)</td>
<td><cite>&#8211;genIndels=1</cite></td>
<td>If set to 1, Platypus will call Indels. If set to 0 Platypus will not call Indels</td>
</tr>
<tr class="row-even"><td>&#8211;minBaseQual</td>
<td>Integer</td>
<td>20</td>
<td><cite>&#8211;minBaseQual=25</cite></td>
<td>Only bases with base quality &gt;= this value will be examined when generating the initial list of SNP candidates</td>
</tr>
<tr class="row-odd"><td>&#8211;minMapQual</td>
<td>Integer</td>
<td>20</td>
<td><cite>&#8211;minMapQual=25</cite></td>
<td>Only bases with base quality &gt;= this value will be examined when generating the initial list of SNP candidates</td>
</tr>
<tr class="row-even"><td>&#8211;maxVariants</td>
<td>Integer</td>
<td>8</td>
<td><cite>&#8211;maxVariants=12</cite></td>
<td>A limit on the number of variant candidates that Platypus will consider in any single window. For more detals see <a class="reference internal" href="#filtering-variant-candidates"><span class="std std-ref">Filtering variant candidates</span></a> for more details</td>
</tr>
<tr class="row-odd"><td>&#8211;maxHaplotypes</td>
<td>Integer</td>
<td>256</td>
<td><cite>&#8211;maxHaplotypes=500</cite></td>
<td>A limit on the number of haplotpyes that Platypus will consider in any single window. For more details see <a class="reference internal" href="#generating-haplotypes"><span class="std std-ref">Generating haplotypes</span></a> for more details</td>
</tr>
<tr class="row-even"><td>&#8211;minReads</td>
<td>Integer</td>
<td>1</td>
<td><cite>&#8211;minReads=10</cite></td>
<td>There must be &gt;= minReads reads with mapping quality &gt;= minMapQual supporting any variant for it to be considered. For SNPs the base qualities must be &gt;= minBaseQual</td>
</tr>
<tr class="row-odd"><td>&#8211;maxSize</td>
<td>Integer</td>
<td>250</td>
<td><cite>&#8211;maxSize=250</cite></td>
<td>An upper limit on the size of indels that Platypus will call</td>
</tr>
<tr class="row-even"><td>&#8211;getVariantsFromBAMs</td>
<td>Boolean</td>
<td>1</td>
<td><cite>&#8211;getVariantsFromBAMs=1</cite></td>
<td>If set to 0 then Playpus will not scan the BAM for variant candidates. This only makes sense if there are candidates coming from elsewhere, e.g. an input VCF of variants</td>
</tr>
<tr class="row-odd"><td>&#8211;strandFilter</td>
<td>Boolean</td>
<td>0</td>
<td><cite>&#8211;strandFilter=1</cite></td>
<td>If set to 1 then Platypus will require at least 1 read on each of the forward and reverse strands in order to consider a variant candidate</td>
</tr>
<tr class="row-even"><td>&#8211;minFlank</td>
<td>Integer</td>
<td>3</td>
<td><cite>&#8211;minFlank=10</cite></td>
<td>Any indel candidate in the BAM must have &gt;= minFlank reference-matching bases on either side of it to be considered</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="variant-filtering-options">
<h2>4.4. Variant filtering options<a class="headerlink" href="#variant-filtering-options" title="Permalink to this headline">¶</a></h2>
<p>The VCF output of Platypus contains a number of metrics and flags that can be used to filter the variant calls. The following options can be used to control the filters.</p>
<table border="1" class="docutils">
<colgroup>
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Option</th>
<th class="head">Type</th>
<th class="head">Default Value</th>
<th class="head">Example Usage</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&#8211;minPosterior</td>
<td>Float</td>
<td>5</td>
<td><cite>&#8211;minPosterior=10</cite></td>
<td>Only variant calls wth a posterior / QUAL value &gt;= this will be output to the VCF. This value is a <a class="reference external" href="https://en.wikipedia.org/wiki/Phred_quality_score">Phred score</a>.</td>
</tr>
<tr class="row-odd"><td>&#8211;sbThreshold</td>
<td>Float</td>
<td>0.01</td>
<td><cite>&#8211;sbThreshold=0.02</cite></td>
<td>P-value threshold for the strand-bias filter. Variants with a strand-bias p-value &lt; this threshold will be flagged as <cite>sb</cite> in the VCF filter column. See <a class="reference internal" href="#filtering-variant-candidates"><span class="std std-ref">Filtering variant candidatess</span></a> for more details.</td>
</tr>
<tr class="row-even"><td>&#8211;abThreshold</td>
<td>Float</td>
<td>0.001</td>
<td><cite>&#8211;abThreshold=0.002</cite></td>
<td>P-value threshold for the allele-bias filter. Variants with an allele-bias p-value &lt; this threshold will be flagged as <cite>ab</cite> in the VCF filter column. See <a class="reference internal" href="#filtering-variant-candidates"><span class="std std-ref">Filtering variant candidates</span></a> for more details.</td>
</tr>
<tr class="row-odd"><td>&#8211;minPosterior</td>
<td>Float</td>
<td>5</td>
<td><cite>&#8211;minPosterior=10</cite></td>
<td>Only variant calls wth a posterior / QUAL value &gt;= this will be output to the VCF. This value is a <a class="reference external" href="https://en.wikipedia.org/wiki/Phred_quality_score">Phred score</a>.</td>
</tr>
<tr class="row-even"><td>&#8211;badReadsWindow</td>
<td>Integer</td>
<td>11</td>
<td><cite>&#8211;badReadsWindow=7</cite></td>
<td>Size of window around variant location to check for low quality bases. If the median minimum quality in this window is &lt; badReadsThreshold then the variant will be filtered with the flag &#8216;badReads&#8217; in the output VCF. See <a class="reference internal" href="#filtering-variant-candidates"><span class="std std-ref">Filtering variant candidates</span></a> for more details.</td>
</tr>
<tr class="row-odd"><td>&#8211;badReadsThreshold</td>
<td>Integer</td>
<td>15</td>
<td><cite>&#8211;badReadsThreshold=20</cite></td>
<td>Threshold for the bad reads filter. See <a class="reference internal" href="#filtering-variant-candidates"><span class="std std-ref">Filtering variant candidates</span></a> for more details.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="miscellaneous-options">
<h2>4.5. Miscellaneous options<a class="headerlink" href="#miscellaneous-options" title="Permalink to this headline">¶</a></h2>
<table border="1" class="docutils">
<colgroup>
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Option</th>
<th class="head">Type</th>
<th class="head">Default Value</th>
<th class="head">Example Usage</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>&#8211;source</td>
<td>String</td>
<td>None</td>
<td><cite>&#8211;source=thousand_genomes_snps.vcf.gz</cite></td>
<td>Name of an input VCF file to be used as a source of variant candidates. See <a class="reference internal" href="#supplying-variant-candidates-from-vcf"><span class="std std-ref">Supplying variant candidates from VCF</span></a> for more details</td>
</tr>
<tr class="row-odd"><td>&#8211;freqAsPrior</td>
<td>Boolean</td>
<td>0</td>
<td><cite>&#8211;freqAsPrior=1</cite></td>
<td>Must be used in conjunction with <cite>&#8211;source</cite>. If this is set to 1 then Platypus will use the AF tag in the input VCF as a prior on the variant. See :ref:Supplying variant candidates from VCF&lt;supplying_variant_candidates_from_vcf&gt;` for more details</td>
</tr>
<tr class="row-even"><td>&#8211;parseNCBI</td>
<td>Boolean</td>
<td>0</td>
<td><cite>&#8211;parseNCBI=1</cite></td>
<td>This must be set to 1 if the reference file has sequence names in the NCBI format</td>
</tr>
<tr class="row-odd"><td>&#8211;printVarsAndExit</td>
<td>Boolean</td>
<td>0</td>
<td><cite>&#8211;printVarsAndExit=1</cite></td>
<td>If set to 1, Platypus will simply print a list of all variant candidates in the calling regions, and exit without making calls</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="deprecated-command-line-options">
<h2>4.6. Deprecated command-line options<a class="headerlink" href="#deprecated-command-line-options" title="Permalink to this headline">¶</a></h2>
<p>The following command-line options are deprecated, and should not be used.</p>
<table border="1" class="docutils">
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Option</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>-n &#8211;nIndividuals</td>
<td>Was used to set the number of individuals in the input BAM files. Now each BAM file is assumed to contain data from only one individual</td>
</tr>
<tr class="row-odd"><td>-p &#8211;ploidy</td>
<td>Was used to set the ploidy of the samples in the BAM files. Now this is fixed at 2.</td>
</tr>
<tr class="row-even"><td>&#8211;dataType</td>
<td>Was used to distinguish between individual, trio and pooled sequencing datasets</td>
</tr>
<tr class="row-odd"><td>&#8211;maxReadLength</td>
<td>Was used to limit the size of reads that Platypus would process</td>
</tr>
<tr class="row-even"><td>&#8211;labels</td>
<td>Was used to specify how to convert BAM file names into the VCF sample column headers</td>
</tr>
<tr class="row-odd"><td>&#8211;callOnlyIndels</td>
<td>If set to 1 then Platypus will only call variants in regions where there are indel candidates</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="specifying-calling-regions">
<span id="id3"></span><h2>4.7. Specifying calling regions<a class="headerlink" href="#specifying-calling-regions" title="Permalink to this headline">¶</a></h2>
<p>It often useful to call variants on only a subset of the genome, e.g. a particular gene or exon. Platypus supports this mode through the <cite>&#8211;regions</cite> option. This option can be used in several ways:</p>
<blockquote>
<div><ul class="simple">
<li>A comma-separated list of chromosome:start-end coordinates can be given, e.g. <cite>&#8211;regions=chr1:0-100,chr2:300-400,chrX:1000-20000</cite>. Platypus will search for variants in just these regions</li>
<li>A comma-separated list of chromosomes can be given, e.g. <cite>&#8211;regions=chr1,chr2,chr20,chrY</cite>. Platypus will search for variants in just these chromosomes, and the lengths of the chromosomes will be taken from either the BAM file header or the reference FASTA file index</li>
<li>A text file name may be specified, e.g. <cite>&#8211;regions=my_regions.txt</cite>. This file must contain only lines with the format <cite>chrom:start-end</cite>, e.g. <cite>chr1:1000-20000</cite> with one region per line.</li>
<li>If no regions are specified then Platypus will call variants across the whole genome. It will check the BAM header file and the reference FASTA file to determine the list of chromosomes and other contigs, as well as their lengths</li>
</ul>
</div></blockquote>
</div>
</div>
<div class="section" id="description-of-variant-calling-algorithm">
<h1>5. Description of Variant Calling Algorithm<a class="headerlink" href="#description-of-variant-calling-algorithm" title="Permalink to this headline">¶</a></h1>
<p>Platypus is designed to call variants on BAM files containing aligned reads generated from the Illumina GAII and Hi-Seq sequencing machines. Platypus calls variants in small haplotype <em>blocks</em> or <em>windows</em>, identifying SNPs and short indels simultaneously. The variant calling algorithm has the following main parts:</p>
<blockquote>
<div><ul class="simple">
<li>Read processing</li>
<li>Construction of variant candidates</li>
<li>Windowing</li>
<li>Haplotype generation</li>
<li>Haplotype likelihood computation</li>
<li>Haplotype frequency estimation</li>
<li>Calling variants and genotypes</li>
<li>Variant filtering</li>
</ul>
</div></blockquote>
<div class="section" id="read-processing">
<h2>5.1. Read processing<a class="headerlink" href="#read-processing" title="Permalink to this headline">¶</a></h2>
<p>Platypus takes as input one or more BAM files containing reads mapped to a reference genome, and an indexed FASTA file of the reference sequence. The BAM data are processed in chunks, sorted by the mapping position in the reference sequence of the reads. To reduce disk load the algorithm first loads all reads mapping to a fairly large region (default 1 mb; option bufferSize) into memory. Each BAM file is assumed to contain a single sample. While this approach works well for modest numbers of samples, it does not scale to large sample sizes (~1000 and up). The reason is that in order to fit all reads into memory, the region size needs to shrink as the sample size increases, increasing the number of region load operations.</p>
<p>After the reads are read into memory, a de‐duplication step removes reads that likely represent PCR copies of the same molecule, as base‐calling or indel errors in these may otherwise be mistaken for real variants. In addition, reads having fewer than 20 high quality bases (option minGoodQualBases), defined as those with a base quality score of at least 20 (option minBaseQual), are also discarded.</p>
<p>By default, reads that have low mapping quality scores (below 20; option minMapQual) are also discarded.</p>
</div>
<div class="section" id="construction-of-candidate-variants">
<span id="generating-variants"></span><h2>5.2. Construction of candidate variants<a class="headerlink" href="#construction-of-candidate-variants" title="Permalink to this headline">¶</a></h2>
<p>Two sources contribute to the list of candidate variants that are considered by the algorithm:</p>
<ul class="simple">
<li>Single‐nucleotide variants (SNVs) and short indels as reported in the CIGAR string and read sequences of the BAM files</li>
<li>Any auxiliary variants provided from a (bgzip‐compressed, tabix‐indexed) VCF file.</li>
</ul>
<p>Each of these sources are optional and controlled by the user (options: getVariantsFromBAMs; source); e.g. Platypus can be made to genotype only a list of specific alleles fed from a source VCF, while for a normal variant calling run, only variants from read alignments are used. Candidate indel variants that occur closer than 3 bp (configurable with option minFlank) to either of the read’s edges are discarded.</p>
<div class="section" id="candidate-variants-from-vcf-files">
<h3>5.2.1. Candidate variants from VCF files<a class="headerlink" href="#candidate-variants-from-vcf-files" title="Permalink to this headline">¶</a></h3>
<p>When a VCF file is provided as input, Platypus collects all variants listed in it,independent of genotype or filter status, and adds these to the list of candidate variants.</p>
</div>
<div class="section" id="priors-on-candidates">
<h3>5.2.2. Priors on candidates<a class="headerlink" href="#priors-on-candidates" title="Permalink to this headline">¶</a></h3>
<p>Platypus assigns priors at the candidate variant generation stage. The following priors are used:</p>
<ul class="simple">
<li>SNPs: 0.33×10‐3. This slightly conservative for multi-sample calling.</li>
<li>Indels (pure insertion or deletion of sequence): we use a model that estimates the local indel rate from local sequence similarity as described in ref. 5, and Supplementary Information section 8. This assigns for example a prior of 4×10‐4 to 1‐bp indels in a homopolymer context of length 4, and a prior of 0.6×10‐2 to 1‐bp indels in a homopolymer context of length 10. Deletions in complex contexts (homopolymer run length of 3 or less) are assigned a prior of 5×10‐5×LI(n), corresponding to one deletion in complex sequence for every 20,000 bp (see 5), and an indel length distribution LI(n), for whereLI(n)=0.25×(3/4)n.Similarly,insertionsareassignedapriorof5×10‐6× LI(n),to reflect the fact that insertions in complex sequence are about 10x less frequent than deletions (see 5). The indel length distribution LI(n) penalizes long indels too strongly; to ensure that long indels are called, we capped the prior from below at a value of 10‐10.</li>
</ul>
</div>
</div>
<div class="section" id="windowing">
<h2>5.3. Windowing<a class="headerlink" href="#windowing" title="Permalink to this headline">¶</a></h2>
<p>Platypus calls alleles and thereby variants in small genomic windows. This section explains how these windows are constructed.
First, any length‐changing variant is left‐aligned, and the set of candidates is filtered for low support. Candidates are kept if they are supported by at least 2 (option: minReads) reads with bases of quality 20 (option: minBaseQual) or above. Candidates from the assembly stage are kept if the accumulated base quality exceeds 40 (minReads*minBaseQual) for each of the variant’s nucleotides. Next, the candidates are sorted by position, and grouped by position. Groups are subsequently merged if they contain overlapping variants.
Platypus then constructs windows by combining groups containing variants that are within 15 bp of each other. If, as a result, the window has become longer than the read length, or the total number of candidate variants within the window exceeds 8, the window is split into smaller windows. If the window is not skipped, but contains more than 8 variants, Platypus prioritizes candidates from assembly, and further prioritizes candidates by read support. In this way the maximum number of candidate variants considered per window is limited to 8 (option maxVariants). For certain designs, including those with large numbers of samples, and those targeting highly diverse regions, it will be beneficial to increase this threshold.</p>
</div>
<div class="section" id="haplotype-generation">
<span id="generating-haplotypes"></span><h2>5.4. Haplotype generation<a class="headerlink" href="#haplotype-generation" title="Permalink to this headline">¶</a></h2>
<p>The most straightforward way to build haplotypes out of a set of candidate variants is to generate all combinations of candidate variants, and apply each set of variants to the reference sequence. For n variants, this results in 2n haplotypes (disregarding excluded combinations due to overlap). For small n this is acceptable, but the procedure results in excessive runtime in a minority of windows with many candidate variants. To address this, we implemented an alternative stepwise prioritization algorithm that is applied for larger n, and that keeps the number of haplotypes that need to be considered under control. With default settings, which allows a maximum of 8 variants in a single window, this algorithm is not used, and all possible combinations are considered.
If the total number of haplotypes that must be considered exceeds a specified limit (the default is 256, or 28), we first pick the variant that is most strongly supported by the data, and build a haplotype from this variant. This haplotype goes into a set H. Next, Platypus goes through all remaining variants, and iteratively adds all combinations of existing haplotypes with that variant to H. Every time a new haplotype h is added, the set of all haplotypes currently in H is ranked by the maximum of likelihoods of the heterozygote (r, h) genotype across all samples, where r denotes the reference. At each step the top K haplotypes are kept, where K=256 by default; this number is configurable (option: maxHaplotypes). The intuition behind this approach is that in most cases, reads either support the reference or an alternative haplotype, and the likelihood of (r,h) is a reasonable proxy of the true marginalized likelihood. In cases where the true genotype is heterozygous non‐reference (h1,h2), each candidate haplotype that approximates h1 is penalized equally by forcing h2 to be explained by r, so that the true h1 has a good chance of making it in the list. In the event of duplicate haplotypes being produced, i.e. the same exact sequence is produced by more than one combination of variants, the combination with the highest overall prior probability is used.</p>
</div>
<div class="section" id="calculation-of-haplotype-likelihoods">
<h2>5.5. Calculation of haplotype likelihoods<a class="headerlink" href="#calculation-of-haplotype-likelihoods" title="Permalink to this headline">¶</a></h2>
<p>This section, together with the following sections, explains how the genotype likelihoods are calculated.
The first step in the calculation of the genotype likelihood, is the calculation of the haplotype likelihood p( r | h ), where r and h denote read and haplotype respectively. These are calculated by aligning a read to the haplotype sequence. The underlying model is a hidden Markov model (HMM), and the likelihood of a read given a haplotype can be calculated using the Forward algorithm. Because this likelihood calculation needs to be performed many times, we used the Viterbi algorithm instead, which admits a more efficient implementation, and which calculates the probability of the most likely path through the HMM, i.e. the probability of the most likely alignment given the data, rather than the likelihood which would marginalize over all possible alignments. In practice, and with the proper choice of gap priors, the Viterbi algorithm is a good practical approximation of the full Forward algorithm.
The alignment algorithm includes models for base mismatches and indel errors. Mismatches are scored by adding up the Phred quality scores of mismatching bases. The likelihood of indel errors is modeled using position‐dependent gap‐opening Phred scores. These are pre‐ calculated based on the propensity of indel errors to occur given the (reference) sequence context. This model, the same as is used by Dindel2 is a simplified version of the model for the indel prior, and considers homopolymers only, rather than homopolymers and more general tandem repeats. Note that the alignment algorithm does not model SNPs and indel mutations, since the haplotype h represents the hypothesized true sequence.</p>
</div>
<div class="section" id="estimation-of-haplotype-frequencies">
<h2>5.6. Estimation of haplotype frequencies<a class="headerlink" href="#estimation-of-haplotype-frequencies" title="Permalink to this headline">¶</a></h2>
<p>After P( r | h ) is calculated for all combinations of reads and haplotypes, an Expectation‐ Maximization (EM) algorithm is run to estimate the frequency of each haplotype h1,...,ha, under a diploid genotype model:</p>
<p>EQUATION</p>
<p>Here, fi denotes the frequency of haplotype hi in the population; a is the number of alleles considered, R and Rs denote the set of all reads, and reads from sample s respectively, and the sum over haplotypes extends over all ordered pairs (i,j), i.e. genotypes. In the formula above we implicitly integrate out the latent variable that determines the two haplotypes of the sample s. This formula holds for both heterozygous and homozygous genotypes.</p>
</div>
<div class="section" id="calling-variants-and-genotypes">
<h2>5.7. Calling variants and genotypes<a class="headerlink" href="#calling-variants-and-genotypes" title="Permalink to this headline">¶</a></h2>
<p>The posterior support for any variant is computed by comparing the likelihood of the data given all haplotypes, and the likelihood given only those haplotypes that do not include a particular variant, i.e., the likelihood in a nested model where the frequencies of haplotypes that do not include the variant are fixed to 0. For the latter model, the frequencies not fixed to 0 are scaled up to account for the estimated frequency of the excluded haplotypes:</p>
<p>EQUATION</p>
<p>where P(v) is the prior probability of observing variant v, Iv is the set of haplotype indices i for which hi does not contain v, and 𝐹! =   !∈!! 𝑓!. The likelihood of reads given haplotypes and their frequencies is computed as</p>
<p>EQUATION</p>
<p>Variants are called when their posterior support exceeds a threshold (by default Phred score 5), using these frequencies as a prior.</p>
<p>Genotype likelihoods for a particular variant are calculated by marginalizing over the genotypes at other variant sites within the window being considered. The best likelihood is reported as a genotype call, and the posterior for this call is calculated in the usual way (as prior times likelihood of the call, divided by the sum of prior times likelihood over all genotypes considered), and reported as a “genotype quality Phred score” (the Phred‐scaled probability of the call being wrong, or 10 times negative 10‐log of one minus the posterior) in the per‐sample GQ field. Using the maximum‐likelihood estimates of haplotype frequencies estimated from the data itself as priors when calling haplotypes and variants works well, but tends to bias genotype calls particularly for small pedigrees and single samples. To address this, we replace the estimated frequencies by a flat prior when calling genotypes if the number of samples is below 25. This also affects the reported genotype quality, but does not affect the reported genotype likelihoods.</p>
</div>
<div class="section" id="filtering">
<span id="filtering-variant-candidates"></span><h2>5.8. Filtering<a class="headerlink" href="#filtering" title="Permalink to this headline">¶</a></h2>
<p>Platypus will only output variants with a phred-scaled posterior probability &gt; 5. In addition, a number of soft filters are applied on a per‐variant basis, with corresponding flags in the <cite>FILTER</cite> column of the output VCF.:</p>
<div class="section" id="allele-bias-ab">
<h3>5.8.1. Allele bias (ab)<a class="headerlink" href="#allele-bias-ab" title="Permalink to this headline">¶</a></h3>
<p>The allele bias filter identifies variants that show support in too few reads compared to the expectation under heterozygous segregation in a diploid organism. Specifically, it rejects variants if (i) the fraction of reads supporting the variant allele is less than the minimum of 0.5 and a user‐specified threshold frequency (default 20%; configurable via the –minVarFreq option); and (ii) the p value under a binomial model with a Beta prior is less than 0.001. We used a Beta prior to account for small biases which may exist even for good calls, for instance due to mapping or PCR biases. The parameters for the Beta distribution are fixed at α=20 and β=20.</p>
</div>
<div class="section" id="strand-bias-sb">
<h3>5.8.2. Strand bias (sb)<a class="headerlink" href="#strand-bias-sb" title="Permalink to this headline">¶</a></h3>
<p>The strand bias filter identifies variants whose support is skewed in terms of reads mapping to the forward and reverse strands, relative to the distribution seen in all reads. We use the distribution seen overall, rather than say a binomial distribution centered around a fraction of 0.5, because certain experimental designs can give rise to bona fide strand biases. Examples include exon capture, and mapping biases due to the existence of an anchoring point to one end of the sequence, but not the other.
Specifically, the reads supporting the variant are tested against a Beta‐binomial distribution with parameters α and β such that that smallest of these parameters is 20, and the parameters are such that the mean of the distribution equals the ratio observed in all reads. Variants are accepted if the p value exceeds 0.001.</p>
</div>
<div class="section" id="bad-reads-badreads">
<h3>5.8.3. Bad reads (badReads)<a class="headerlink" href="#bad-reads-badreads" title="Permalink to this headline">¶</a></h3>
<p>This filter triggers when across reads supporting a variant, the median of the minimum base quality close to the focal site (default 7 bp either side, configurable using ‐‐badReadsWindow) is too low (default 15 or less, configurable using ‐‐badReadsThreshold); this identifies systematic local sequencing issues causing an excess of read errors, which are not always accurately reflected in the read quality scores. It also triggers when more than a fraction of reads are filtered out for the candidate generation stage; the default for this is 0.7 (configurable using – filteredReadsFrac).</p>
</div>
<div class="section" id="homopolymers-filter-hp10">
<h3>5.8.4. Homopolymers filter (hp10)<a class="headerlink" href="#homopolymers-filter-hp10" title="Permalink to this headline">¶</a></h3>
<p>Polymerase slippage in low‐complexity regions are a known cause of spurious indel calls, and in the alignment model we account for a higher incidence of these errors. However, in certain instances, particularly in long homopolymers, our error model does not compensate for the high rate of sequencing errors, and spurious indel calls may result. To avoid a high false-positive rate in such regions, Platypus flags any varian call as suspicious if the haplotype containing the variant has a homopolymer of length 10 or greater that crosses the variant site.</p>
</div>
</div>
<div class="section" id="supplying-variant-candidates-from-vcf">
<span id="id4"></span><h2>5.9. Supplying variant candidates from VCF<a class="headerlink" href="#supplying-variant-candidates-from-vcf" title="Permalink to this headline">¶</a></h2>
<p>Blah</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">1. Introduction</a></li>
<li><a class="reference internal" href="#installation">2. Installation</a><ul>
<li><a class="reference internal" href="#dependencies">2.1. Dependencies</a></li>
</ul>
</li>
<li><a class="reference internal" href="#running-platypus">3. Running Platypus</a></li>
<li><a class="reference internal" href="#command-line-options">4. Command Line Options</a><ul>
<li><a class="reference internal" href="#basic-options">4.1. Basic options</a></li>
<li><a class="reference internal" href="#bam-data-processing-and-filtering-options">4.2. BAM Data processing and filtering options</a></li>
<li><a class="reference internal" href="#variant-calling-options">4.3. Variant calling options</a></li>
<li><a class="reference internal" href="#variant-filtering-options">4.4. Variant filtering options</a></li>
<li><a class="reference internal" href="#miscellaneous-options">4.5. Miscellaneous options</a></li>
<li><a class="reference internal" href="#deprecated-command-line-options">4.6. Deprecated command-line options</a></li>
<li><a class="reference internal" href="#specifying-calling-regions">4.7. Specifying calling regions</a></li>
</ul>
</li>
<li><a class="reference internal" href="#description-of-variant-calling-algorithm">5. Description of Variant Calling Algorithm</a><ul>
<li><a class="reference internal" href="#read-processing">5.1. Read processing</a></li>
<li><a class="reference internal" href="#construction-of-candidate-variants">5.2. Construction of candidate variants</a><ul>
<li><a class="reference internal" href="#candidate-variants-from-vcf-files">5.2.1. Candidate variants from VCF files</a></li>
<li><a class="reference internal" href="#priors-on-candidates">5.2.2. Priors on candidates</a></li>
</ul>
</li>
<li><a class="reference internal" href="#windowing">5.3. Windowing</a></li>
<li><a class="reference internal" href="#haplotype-generation">5.4. Haplotype generation</a></li>
<li><a class="reference internal" href="#calculation-of-haplotype-likelihoods">5.5. Calculation of haplotype likelihoods</a></li>
<li><a class="reference internal" href="#estimation-of-haplotype-frequencies">5.6. Estimation of haplotype frequencies</a></li>
<li><a class="reference internal" href="#calling-variants-and-genotypes">5.7. Calling variants and genotypes</a></li>
<li><a class="reference internal" href="#filtering">5.8. Filtering</a><ul>
<li><a class="reference internal" href="#allele-bias-ab">5.8.1. Allele bias (ab)</a></li>
<li><a class="reference internal" href="#strand-bias-sb">5.8.2. Strand bias (sb)</a></li>
<li><a class="reference internal" href="#bad-reads-badreads">5.8.3. Bad reads (badReads)</a></li>
<li><a class="reference internal" href="#homopolymers-filter-hp10">5.8.4. Homopolymers filter (hp10)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#supplying-variant-candidates-from-vcf">5.9. Supplying variant candidates from VCF</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Welcome to Platypus 0.1.5 documentation!</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/documentation.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="index.html" title="Welcome to Platypus 0.1.5 documentation!"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Platypus 0.1.5 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, Andy Rimmer.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.3.
    </div>
  </body>
</html>