<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>1. Introduction &#8212; Platypus 0.1.5 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.5',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Welcome to Platypus 0.1.5 documentation!" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="index.html" title="Welcome to Platypus 0.1.5 documentation!"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Platypus 0.1.5 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="introduction">
<h1>1. Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h1>
<p>Platypus is a tool for identifying small variants (SNPs and indels) in BAM files. This code is a fork of Platypus version 0.1.5, which is used
in OpEx. If you are looking for the most recent version of Platypus, the code is available <a class="reference external" href="https://github.com/andyrimmer/Platypus">here</a>.</p>
</div>
<div class="section" id="installation">
<h1>2. Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p>Stable releases of Platypus can be downloaded from Github <a class="reference external" href="https://github.com/RahmanTeamDevelopment/Platypus/releases">here</a>
in either <code class="docutils literal"><span class="pre">.zip</span></code> or <code class="docutils literal"><span class="pre">.tar.gz</span></code> format. To unpack these run one of the following commands:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">unzip</span> <span class="n">Platypus</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">zip</span>
</pre></div>
</div>
<p>or:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">tar</span> <span class="o">-</span><span class="n">xvzf</span> <span class="n">Platypus</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>and then you can install Platypus with the following commands:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="n">Platypus</span>
<span class="o">./</span><span class="n">install</span><span class="o">.</span><span class="n">sh</span>
</pre></div>
</div>
<p>Platypus uses <code class="docutils literal"><span class="pre">virtualenv</span></code> and <code class="docutils literal"><span class="pre">pip</span></code> to manage all its extra dependencies, which means that it will not clutter up your system by installing
things globally. Everthing it installs will go into a sub-directory in the <code class="docutils literal"><span class="pre">Platypus</span></code> directory (specifically, <code class="docutils literal"><span class="pre">Platypus_X.X.X/env</span></code>). If
you delete Platypus then everything it has installed will also be deleted. Once the installation script has finished successfully,
Platypus is ready for use.</p>
<div class="section" id="dependencies">
<h2>2.1. Dependencies<a class="headerlink" href="#dependencies" title="Permalink to this headline">¶</a></h2>
<p>To install and run Platypus v0.1.5 you will need <a class="reference external" href="https://www.python.org">Python</a> (only version 2.7.X is currently supported),
and <a class="reference external" href="https://virtualenv.pypa.io/en/stable/">Virtualenv</a>.</p>
</div>
</div>
<div class="section" id="running-platypus">
<h1>3. Running Platypus<a class="headerlink" href="#running-platypus" title="Permalink to this headline">¶</a></h1>
<p>Once downloaded and correctly installed, Platypus can be run with the following simple command:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">env</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">platypus</span> <span class="n">callVariants</span> <span class="o">--</span><span class="n">bamFiles</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">bam</span> <span class="o">--</span><span class="n">refFile</span><span class="o">=</span><span class="n">reference</span><span class="o">.</span><span class="n">fa</span> <span class="o">--</span><span class="n">output</span><span class="o">=</span><span class="n">variant_calls</span><span class="o">.</span><span class="n">vcf</span>
</pre></div>
</div>
<p>This is the simplest way to run Platypus, specifying only the names of the input BAM file and input reference file. Variant
calls will then be outupt to a file called <cite>AllVariants.vcf</cite>.</p>
<ul class="simple">
<li>The input bam file should follow the <a class="reference external" href="http://samtools.github.io/hts-specs/SAMv1.pdf">BAM format</a> and should contain mapped reads sorted by genomic coordinate. Platypus has been tested using BAM files produced by <a class="reference external" href="http://www.well.ox.ac.uk/project-stampy">Stampy</a> and <a class="reference external" href="http://bio-bwa.sourceforge.net">BWA</a>, but should also work with other short read aligners. The BAM index file (.bai) should also be present in the same directory; this can be created using <a class="reference external" href="http://www.htslib.org">Samtools</a>.</li>
</ul>
</div>
<div class="section" id="command-line-options">
<h1>4. Command Line Options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h1>
<p>Platypus has a large number of configuration options that can be specified on the command-line. Most of these can be ignored, as they take default values
and are already configured for optimal results. Below is a full descripton of all configuration options.</p>
<table border="1" class="docutils">
<colgroup>
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Option</th>
<th class="head">Type</th>
<th class="head">Default Value</th>
<th class="head">Example Usage</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>-h &#8211;help</td>
<td>NA</td>
<td>NA</td>
<td><cite>&#8211;help</cite></td>
<td>If this option is set then Platypus will print a help message and then exit</td>
</tr>
<tr class="row-odd"><td>&#8211;bamFiles</td>
<td>Comma-separated list of strings</td>
<td>None</td>
<td><cite>&#8211;bamFiles=test1.bam,test2.bam</cite></td>
<td>A list of BAM file names. All the specified BAM files will be searched for variants. The output will go into a single VCF file</td>
</tr>
<tr class="row-even"><td>&#8211;refFile</td>
<td>String</td>
<td>None</td>
<td><cite>&#8211;refFile=human_37.fa</cite></td>
<td>Name of the input reference FASTA file. This file must be indexed (using <cite>Samtools faidx</cite> for example) and must contain the same reference sequence that was used to map the reads in the input BAM file(s)</td>
</tr>
<tr class="row-odd"><td>-o &#8211;output</td>
<td>String</td>
<td>AllVariants.vcf</td>
<td><cite>&#8211;output=variant_calls.vcf</cite></td>
<td>Name of the output VCF file</td>
</tr>
<tr class="row-even"><td>&#8211;regions</td>
<td>Comma-separated list of strings or the name of a text file</td>
<td>None</td>
<td><cite>&#8211;regions=chr1:1000-2000,chr2:2000-4000</cite></td>
<td>Platypus will produce variant calls only in the specified regions. See <a class="reference internal" href="#specifying-calling-regions"><span class="std std-ref">Specifying calling regions</span></a> for a more detailed description</td>
</tr>
<tr class="row-odd"><td>&#8211;verbosity</td>
<td>Integer from 0 to 3</td>
<td>2</td>
<td><cite>&#8211;verbosity=3</cite></td>
<td>Sets the level of output logging for Platypus. Increase this value for more verbose log output</td>
</tr>
<tr class="row-even"><td>&#8211;logFileName</td>
<td>String</td>
<td>log.txt</td>
<td><cite>&#8211;logFileName=platypus_log_file.txt</cite></td>
<td>Name of the output log file.</td>
</tr>
<tr class="row-odd"><td>&#8211;nCPU</td>
<td>Integer</td>
<td>1</td>
<td><cite>&#8211;nCPU=3</cite></td>
<td>Number of processes to use for variant calling. If &gt; 1 then Platypus will run in multiple processes and merge the VCF files into one file at the end</td>
</tr>
</tbody>
</table>
<div class="section" id="deprecated-command-line-options">
<h2>4.1. Deprecated command-line options<a class="headerlink" href="#deprecated-command-line-options" title="Permalink to this headline">¶</a></h2>
<p>The following command-line options are deprecated, and should not be used.</p>
<table border="1" class="docutils">
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Option</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>-n &#8211;nIndividuals</td>
<td>Was used to set the number of individuals in the input BAM files. Now each BAM file is assumed to contain data from only one individual</td>
</tr>
<tr class="row-odd"><td>-p &#8211;ploidy</td>
<td>Was used to set the ploidy of the samples in the BAM files. Now this is fixed at 2.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="specifying-calling-regions">
<span id="id2"></span><h2>4.2. Specifying calling regions<a class="headerlink" href="#specifying-calling-regions" title="Permalink to this headline">¶</a></h2>
<p>It often useful to call variants on only a subset of the genome, e.g. a particular gene or exon. Platypus supports this mode through the <cite>&#8211;regions</cite> option. This option can be used in several ways:</p>
<blockquote>
<div><ul class="simple">
<li>A comma-separated list of chromosome:start-end coordinates can be given, e.g. <cite>&#8211;regions=chr1:0-100,chr2:300-400,chrX:1000-20000</cite>. Platypus will search for variants in just these regions</li>
<li>A comma-separated list of chromosomes can be given, e.g. <cite>&#8211;regions=chr1,chr2,chr20,chrY</cite>. Platypus will search for variants in just these chromosomes, and the lengths of the chromosomes will be taken from either the BAM file header or the reference FASTA file index</li>
<li>A text file name may be specified, e.g. <cite>&#8211;regions=my_regions.txt</cite>. This file must contain only lines with the format <cite>chrom:start-end</cite>, e.g. <cite>chr1:1000-20000</cite> with one region per line.</li>
<li>If no regions are specified then Platypus will call variants across the whole genome. It will check the BAM header file and the reference FASTA file to determine the list of chromosomes and other contigs, as well as their lengths</li>
</ul>
</div></blockquote>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--genSNPs=<var>GENSNPS</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>If set to TRUE (default), SNP candidates will be
considered</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--genIndels=<var>GENINDELS</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>If set to TRUE (default), Indel candidates will be
considered</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-n <var>NIND</var></span>, <span class="option">--nIndividuals=<var>NIND</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Number of Individuals sequenced</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-p <var>PLOIDY</var></span>, <span class="option">--ploidy=<var>PLOIDY</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Number of copief of each chromosome per individual
same directory
just list of chr, or nothing</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--processRegionSize=<var>PROCESSREGIONSIZE</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Size of region to use in each process. Chromosomes
will be broken up into regions of this size and run in
separate processes (or consecutively, if nCPU == 1)</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--bufferSize=<var>BUFFERSIZE</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Data will be buffered in regions of this size</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--minReads=<var>MINREADS</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Minimum required number of reads in window</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--maxHaplotypes=<var>MAXHAPLOTYPES</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Maximium haplotypes to consider in a given window</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--maxVariants=<var>MAXVARIANTS</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Maximium variants to consider in a given window</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--maxReads=<var>MAXREADS</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Maximium coverage in window</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--maxSize=<var>MAXSIZE</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Largest indel to consider</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--minMapQual=<var>MINMAPQUAL</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Minimum mapping quality of read. Any reads with map
qual below this are ignored</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--maxBadQualBases=<var>MAXBADQUALBASES</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Max number of bases per read that can have base-
quality &lt;= 20</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--minGoodQualBases=<var>MINGOODQUALBASES</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Minimum number of bases per read that must have base-
quality &gt;= 20</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--minBaseQual=<var>MINBASEQUAL</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Minimum allowed base-calling quality. Any bases with
qual below this are ignored in SNP-calling</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--maxReadLength=<var>RLEN</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Maximum read length</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--minFlank=<var>MINFLANK</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Flank size for indel candidates</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--labels=<var>LABELS</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>regex to extract names from filenames</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--source=<var>SOURCEFILE</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>vcf file(s) to get candidates from</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--freqAsPrior=<var>FREQASPRIOR</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>If 1, use the frequency of input variants as a prior</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--dataType=<var>DATATYPE</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Are we looking at individual,trio,population, or pool
data?</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--getVariantsFromBAMs=<var>GETVARIANTSFROMBAMS</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>If set to TRUE (default), variant candidates will be
generated from BAMs as well as any other inputs</td></tr>
</tbody>
</table>
<p>&#8211;parseNCBI=PARSENCBI
&#8211;callOnlyIndels=CALLONLYINDELS</p>
<blockquote>
<div>If set to TRUE (default), only windows containing
Indel candidates will be considered</div></blockquote>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--minPosterior=<var>MINPOSTERIOR</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Only variants with posterior &gt;= this will be outpu to
the VCF. Value is a Phred-score.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--strandFilter=<var>STRANDFILTER</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>If set to TRUE only variants occuring at least once on
each strand will be considered.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--sbThreshold=<var>SBTHRESHOLD</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>P-value for strand-bias filtering..</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--abThreshold=<var>ABTHRESHOLD</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>P-value for allele-bias filtering..</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--printVarsAndExit=<var>PRINTVARSANDEXIT</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>If 1, print a list of variant candidates, and exit.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--badReadsWindow=<var>BADREADSWINDOW</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Size of window around variant to look for low-quality
bases.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--badReadsThreshold=<var>BADREADSTHRESHOLD</var></span></kbd></td>
</tr>
<tr><td>&#160;</td><td>Variants where the median minimum quality in a window
of badReadsWindow around the variant position falls
below this value will be filtered with the flag
&#8216;badReads&#8217;</td></tr>
</tbody>
</table>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">1. Introduction</a></li>
<li><a class="reference internal" href="#installation">2. Installation</a><ul>
<li><a class="reference internal" href="#dependencies">2.1. Dependencies</a></li>
</ul>
</li>
<li><a class="reference internal" href="#running-platypus">3. Running Platypus</a></li>
<li><a class="reference internal" href="#command-line-options">4. Command Line Options</a><ul>
<li><a class="reference internal" href="#deprecated-command-line-options">4.1. Deprecated command-line options</a></li>
<li><a class="reference internal" href="#specifying-calling-regions">4.2. Specifying calling regions</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Welcome to Platypus 0.1.5 documentation!</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/documentation.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="index.html" title="Welcome to Platypus 0.1.5 documentation!"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Platypus 0.1.5 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, Andy Rimmer.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.3.
    </div>
  </body>
</html>